vmproxy/pkg/proxy/proxy.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2022 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package proxy provides http reverse proxy functionality
+package proxy
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Config defines options for starting proxy
+type Config struct {
+	// Name of the header to check for filters. Case insensitive.
+	HeaderName string
+	// Address the proxy is listening on
+	ListenAddress string
+	// Target URL to forward requests to
+	TargetURL *url.URL
+}
+
+// StartProxy starts proxy which adds extra filters based on configuration.
+func StartProxy(cfg Config) {
+	logrus.Infof("Starting to proxy at http://%s to %s", cfg.ListenAddress, cfg.TargetURL.String())
+
+	err := http.ListenAndServe(cfg.ListenAddress, getHandler(cfg))
+	logrus.Error(err)
+}
+
+func getHandler(cfg Config) http.HandlerFunc {
+	rProxy := &httputil.ReverseProxy{
+		Director: director(cfg.TargetURL, cfg.HeaderName),
+	}
+
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		if failOnInvalidHeader(rw, req, cfg.HeaderName) {
+			return
+		}
+
+		rProxy.ServeHTTP(rw, req)
+	})
+}
+
+func failOnInvalidHeader(rw http.ResponseWriter, req *http.Request, headerName string) bool {
+	if filters := req.Header.Get(headerName); filters != "" {
+		if _, err := parseFilters(filters); err != nil {
+			rw.Header().Set("Content-Type", "text/plain; charset=utf-8")
+			rw.WriteHeader(http.StatusPreconditionFailed)
+			io.WriteString(rw, fmt.Sprintf("Failed to parse %s header", headerName))
+			return true
+		}
+	}
+
+	return false
+}
+
+func director(target *url.URL, headerName string) func(*http.Request) {
+	return func(req *http.Request) {
+		logrus.Infof("%s: %s", req.Method, req.URL)
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+
+		// Replace extra filters if present
+		if filters := req.Header.Get(headerName); filters != "" {
+			q := req.URL.Query()
+			q.Del("extra_filters")
+
+			if parsed, _ := parseFilters(filters); parsed != nil {
+				for _, f := range parsed {
+					q.Add("extra_filters", f)
```

Comment:
accordingly to https://github.com/VictoriaMetrics/VictoriaMetrics/#prometheus-querying-api-enhancements it shall be 
```suggestion
					q.Add("extra_filters[]", f)
```

Reply:
and the value shall be like:
```
{env=~"prod|staging",user="xyz"}
```

but not just array of strings

Reply:
and one more - need to understand how several filters with condition "OR" between them shall be added to query

Reply:
vmproxy does not validate the format of the filters. That is left up to VictoriaMetrics. It just passes the provided filters.

Reply:
> and one more - need to understand how several filters with condition "OR" between them shall be added to query

By sending the query parameter multiple times.